mnemonics/src/main/java/org/kethereum/bip32/ExtendedKey.kt

```diff
@@ -0,0 +1,257 @@
+        var result = keyPair.hashCode()
+        result = 31 * result + Arrays.hashCode(chainCode)
+        result = 31 * result + depth
+        result = 31 * result + parentFingerprint
+        result = 31 * result + sequence
+        return result
+    }
+
+    fun serialize(publicKeyOnly: Boolean = false): String {
+        val out = ByteBuffer.allocate(Companion.EXTENDED_KEY_SIZE)
+        try {
+            if (publicKeyOnly || keyPair.privateKey == BigInteger.ZERO) {
+                out.put(xpub)
+            } else {
+                out.put(xprv)
+            }
+            out.put(depth)
+            out.putInt(parentFingerprint)
+            out.putInt(sequence)
+            out.put(chainCode)
+            if (publicKeyOnly || keyPair.privateKey == BigInteger.ZERO) {
+                out.put(getCompressedPublicKey(keyPair))
+            } else {
+                out.put(0x00)
+                out.put(keyPair.privateKey.toBytesPadded(PRIVATE_KEY_SIZE))
+            }
+        } catch (e: IOException) {
+        }
+
+        return out.array().encodeToBase58WithChecksum()
+    }
+
+    companion object {
+
+        private val BITCOIN_SEED = "Bitcoin seed".toByteArray()
+        private val CHAINCODE_SIZE = PRIVATE_KEY_SIZE
+        private val COMPRESSED_PUBLIC_KEY_SIZE = PRIVATE_KEY_SIZE + 1
+        internal val xprv = byteArrayOf(0x04, 0x88.toByte(), 0xAD.toByte(), 0xE4.toByte())
+        internal val xpub = byteArrayOf(0x04, 0x88.toByte(), 0xB2.toByte(), 0x1E.toByte())
+
+        fun createFromSeed(seed: ByteArray, publicKeyOnly: Boolean = false): ExtendedKey {
+            try {
+                val mac = Mac.getInstance("HmacSHA512")
+                val seedKey = SecretKeySpec(BITCOIN_SEED, "HmacSHA512")
+                mac.init(seedKey)
+                val lr = mac.doFinal(seed)
+                val l = Arrays.copyOfRange(lr, 0, PRIVATE_KEY_SIZE)
+                val r = Arrays.copyOfRange(lr, PRIVATE_KEY_SIZE, PRIVATE_KEY_SIZE + CHAINCODE_SIZE)
+                val m = BigInteger(1, l)
+                if (m >= CURVE.n) {
+                    throw KeyException("Master key creation resulted in a key with higher modulus. Suggest deriving the next increment.")
+                }
+                val keyPair = ECKeyPair.create(l)
+                return if (publicKeyOnly) {
+                    val pubKeyPair = ECKeyPair(BigInteger.ZERO, keyPair.publicKey)
+                    ExtendedKey(pubKeyPair, r, 0, 0, 0)
+                } else {
+                    ExtendedKey(keyPair, r, 0, 0, 0)
+                }
+            } catch (e: NoSuchAlgorithmException) {
+                throw KeyException(e)
+            } catch (e: NoSuchProviderException) {
+                throw KeyException(e)
+            } catch (e: InvalidKeyException) {
+                throw KeyException(e)
+            }
+
+        }
+
+        private val HARDENING_FLAG = 0x80000000.toInt()
+        private fun isHardened(element: Int): Boolean {
```

Comment:
this is WET as this already is in ExtendedKey - lets make it DRY

Reply:
moved the relevant code to the BIP44 module where the hardening flag already existed.